File: src/test/java/com/baulsupp/oksocial/integration/WebServerTest.java
package com.baulsupp.oksocial.integration;

import com.baulsupp.oksocial.Main;
import com.google.common.collect.Lists;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.internal.tls.SslClient;
import okhttp3.mockwebserver.MockResponse;
import okhttp3.mockwebserver.MockWebServer;
import org.junit.Rule;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class WebServerTest {
  @Rule public MockWebServer server = new MockWebServer();
  private Main main = new Main();
  private SslClient sslClient = SslClient.localhost();

  @Test public void httpsRequestInsecure() throws Exception {
    server.useHttps(sslClient.socketFactory, false);
    server.enqueue(new MockResponse().setBody("Isla Sorna"));

    main.arguments = Lists.newArrayList(server.url("/").toString());
    main.allowInsecure = true;

    OkHttpClient client = main.createClientBuilder().build();

    Request.Builder requestBuilder = main.createRequestBuilder();

    Request request = requestBuilder.url(server.url("/")).build();

    try (Response response = client.newCall(request).execute()) {
      assertEquals(200, response.code());
    }
  }
}

File: src/test/kotlin/com/baulsupp/okurl/i9n/WebServerTest.kt
package com.baulsupp.okurl.i9n

import com.baulsupp.schoutput.handler.TestOutputHandler
import com.baulsupp.okurl.Main
import com.baulsupp.okurl.okhttp.localhost
import com.baulsupp.okurl.security.CertificatePin
import kotlinx.coroutines.runBlocking
import okhttp3.Response
import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import javax.net.ssl.SSLHandshakeException
import javax.net.ssl.SSLPeerUnverifiedException
import kotlin.test.assertEquals
import kotlin.test.assertTrue

@Disabled
class WebServerTest {
  @JvmField
  var server = MockWebServer()

  private val main = Main()
  private val output = TestOutputHandler<Response>()

  init {
    main.outputHandler = output
  }

  private val sslClient = localhost()

  @Test
  fun httpsRequestInsecureFails() {
    server.useHttps(sslClient.sslSocketFactory(), false)
    server.enqueue(MockResponse().setBody("Isla Sorna"))

    main.arguments = mutableListOf(server.url("/").toString())

    runBlocking {
      main.run()
    }

    assertEquals(0, output.responses.size)
    assertEquals(1, output.failures.size)
    assertTrue(output.failures[0] is SSLHandshakeException)
  }

  @Test
  fun httpsRequestInsecure() {
    server.useHttps(sslClient.sslSocketFactory(), false)
    server.enqueue(MockResponse().setBody("Isla Sorna"))

    main.arguments = mutableListOf(server.url("/").toString())
    main.insecureHost = listOf(server.hostName)

    runBlocking {
      main.run()
    }

    assertEquals(1, output.responses.size)
    assertEquals(200, output.responses[0].code)
  }

  @Test
  @Disabled
  fun httpsRequestSecure() {
    server.useHttps(sslClient.sslSocketFactory(), false)
    server.enqueue(MockResponse().setBody("Isla Sorna"))

    main.arguments = mutableListOf(server.url("/").toString())

    runBlocking {
      main.run()
    }

    assertEquals(1, output.responses.size)
    assertEquals(200, output.responses[0].code)
  }

  @Test
  fun rejectedWithPin() {
    server.useHttps(sslClient.sslSocketFactory(), false)
    server.enqueue(MockResponse().setBody("Isla Sorna"))

    main.arguments = mutableListOf(server.url("/").toString())
    main.certificatePins = listOf(CertificatePin(server.hostName + ":" +
      "sha256/WoiWRyIOVNa9ihaBciRSC7XHjliYS9VwUGOIud4PB18="))
    main.insecureHost = listOf(server.hostName)

    runBlocking {
      main.run()
    }

    assertEquals(0, output.responses.size)
    assertEquals(1, output.failures.size)
    assertTrue(output.failures[0] is SSLPeerUnverifiedException)
  }
}
